File: android/src/main/java/com/appodeal/rnappodeal/RCTAppodealBannerView.java
package com.appodeal.rnappodeal;

import android.app.Activity;
import android.content.Context;
import android.content.res.Resources;
import android.os.Handler;
import android.os.Looper;
import android.util.DisplayMetrics;
import android.util.TypedValue;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.appodeal.ads.Appodeal;
import com.appodeal.ads.BannerCallbacks;
import com.appodeal.ads.BannerView;
import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.ReactContext;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.uimanager.events.RCTEventEmitter;
import com.facebook.react.views.view.ReactViewGroup;

public class RCTAppodealBannerView extends ReactViewGroup implements BannerCallbacks {

    enum BannerSize {PHONE, TABLET}

    private static final String defaultPlacement = "default";
    private static final Handler handler = new Handler(Looper.getMainLooper());

    private BannerSize size = BannerSize.PHONE;
    private String placement = defaultPlacement;

    @Nullable
    private Runnable showRunnable = null;

    private final Runnable measureAndLayout = () -> {
        for (int i = 0; i < getChildCount(); i++) {
            View child = getChildAt(i);
            child.measure(
                    MeasureSpec.makeMeasureSpec(getMeasuredWidth(), MeasureSpec.EXACTLY),
                    MeasureSpec.makeMeasureSpec(getMeasuredHeight(), MeasureSpec.EXACTLY)
            );
            child.layout(0, 0, child.getMeasuredWidth(), child.getMeasuredHeight());
        }
    };

    public RCTAppodealBannerView(Context context) {
        super(context);
    }

    @Override
    public void requestLayout() {
        super.requestLayout();
        post(measureAndLayout);
    }

    public void setAdSize(String adSize) {
        if (adSize.equals("tablet")) {
            this.size = BannerSize.TABLET;
        } else {
            this.size = BannerSize.PHONE;
        }
        cacheAdIfNeeded();
    }

    public BannerSize getSize() {
        return this.size;
    }

    public void setPlacement(String placement) {
        this.placement = placement;
    }

    public String getPlacement() {
        return placement;
    }

    public void showBannerView(@NonNull BannerView bannerView) {
        showRunnable = () -> {
            Activity activity = getReactContext().getCurrentActivity();
            if (activity == null) return;

            Appodeal.set728x90Banners(size == RCTAppodealBannerView.BannerSize.TABLET);

            int height = getEstimatedHeight();
            Resources resources = getReactContext().getResources();
            DisplayMetrics dm = resources.getDisplayMetrics();
            int widthPixels = resources.getDisplayMetrics().widthPixels;
            int heightPixels = dp2px(height, dm);

            LayoutParams bannerLayoutParams = new BannerView.LayoutParams(widthPixels, heightPixels);
            bannerView.setLayoutParams(bannerLayoutParams);
            bannerView.setVisibility(VISIBLE);

            removeAllViews();
            setVisibility(VISIBLE);
            addView(bannerView);

            bannerView.bringToFront();

            String showPlacement = placement == null ? defaultPlacement : placement;
            Appodeal.show(activity, Appodeal.BANNER_VIEW, showPlacement);
        };
        handler.postDelayed(showRunnable, 250L);
    }

    public void hideBannerView(@NonNull BannerView bannerView) {
        removeAllViews();
        ViewGroup parent = (ViewGroup) bannerView.getParent();
        if (parent != null) {
            parent.removeView(bannerView);
        }
        handler.removeCallbacks(showRunnable);
    }

    private void cacheAdIfNeeded() {
        if (!Appodeal.isAutoCacheEnabled(Appodeal.BANNER_VIEW)) {
            Activity activity = getReactContext().getCurrentActivity();
            if (activity != null) {
                Appodeal.cache(activity, Appodeal.BANNER_VIEW);
            }
        }
    }

    private int getEstimatedHeight() {
        if (size == BannerSize.TABLET) {
            return 90;
        }
        return 50;
    }

    private ReactContext getReactContext() {
        return (ReactContext) getContext();
    }

    private RCTEventEmitter getEmitter() {
        return getReactContext().getJSModule(RCTEventEmitter.class);
    }

    private int dp2px(int dp, DisplayMetrics dm) {
        return Math.round(TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, dp, dm));
    }

    @Override
    public void onBannerLoaded(int height, boolean isPrecache) {
        WritableMap params = Arguments.createMap();
        params.putInt("height", height);
        params.putBoolean("isPrecache", isPrecache);
        getEmitter().receiveEvent(getId(), "onBannerLoaded", params);
    }

    @Override
    public void onBannerFailedToLoad() {
        getEmitter().receiveEvent(getId(), "onBannerFailedToLoad", null);
    }

    @Override
    public void onBannerClicked() {
        getEmitter().receiveEvent(getId(), "onBannerClicked", null);
    }

    @Override
    public void onBannerExpired() {
        getEmitter().receiveEvent(getId(), "onBannerExpired", null);
    }

    @Override
    public void onBannerShowFailed() {
    }

    @Override
    public void onBannerShown() {
    }
}

File: android/src/main/java/com/appodeal/rnappodeal/RCTAppodealBannerView.kt
package com.appodeal.rnappodeal

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.util.DisplayMetrics
import android.util.Log
import android.util.TypedValue
import android.view.ViewGroup
import android.widget.FrameLayout
import com.appodeal.ads.Appodeal
import com.appodeal.ads.BannerView
import com.appodeal.rnappodeal.callbacks.RNAppodealEventHandler
import com.appodeal.rnappodeal.constants.BannerEvents
import com.facebook.react.bridge.ReactContext
import com.facebook.react.bridge.WritableMap
import com.facebook.react.uimanager.UIManagerHelper
import com.facebook.react.uimanager.events.Event
import com.facebook.react.uimanager.events.EventDispatcher
import com.facebook.react.views.view.ReactViewGroup
import kotlin.math.roundToInt

class RCTAppodealBannerView(context: Context) : ReactViewGroup(context), RNAppodealEventHandler {

    private val surfaceId: Int by lazy { UIManagerHelper.getSurfaceId(reactContext) }
    private val uiHandler: Handler by lazy { Handler(Looper.getMainLooper()) }

    var placement: String = "default"
        set(value) {
            field = value
            cacheAdIfNeeded()
        }

    var adSize: String = "phone"
        set(value) {
            field = value
            use728x90Banners = value == "tablet"
            cacheAdIfNeeded()
        }

    private var use728x90Banners: Boolean = false
    private var showRunnable: Runnable? = null

    private val measureAndLayout = Runnable {
        for (i in 0 until childCount) {
            val child = getChildAt(i)
            child.measure(
                MeasureSpec.makeMeasureSpec(measuredWidth, MeasureSpec.EXACTLY),
                MeasureSpec.makeMeasureSpec(measuredHeight, MeasureSpec.EXACTLY)
            )
            child.layout(0, 0, child.measuredWidth, child.measuredHeight)
        }
    }

    override fun requestLayout() {
        super.requestLayout()
        post(measureAndLayout)
    }

    fun showBannerView(bannerView: BannerView) {
        val runnable = Runnable {
            val activity = reactContext.currentActivity ?: return@Runnable
            Appodeal.set728x90Banners(use728x90Banners)

            val dpHeight = if (use728x90Banners) 90 else 50
            val heightPx = dp2px(dpHeight, reactContext.resources.displayMetrics)
            val widthPx = reactContext.resources.displayMetrics.widthPixels

            bannerView.layoutParams = FrameLayout.LayoutParams(widthPx, heightPx)
            bannerView.visibility = VISIBLE
            bannerView.descendantFocusability = FOCUS_BLOCK_DESCENDANTS

            removeAllViews()
            this.visibility = VISIBLE
            addView(bannerView)
            bannerView.bringToFront()

            try {
                // Ensure this banner view is brought to front in parent hierarchy
                (parent as? ViewGroup)?.bringChildToFront(this)
            } catch (e: Exception) {
                Log.e("RTCAppodealBanner", "Error bringing banner view to front", e)
            }

            Appodeal.show(activity, Appodeal.BANNER_VIEW, placement)
        }.also { showRunnable = it }
        uiHandler.postDelayed(runnable, 250L)
    }

    fun hideBannerView(bannerView: BannerView) {
        removeAllViews()
        (bannerView.parent as? ViewGroup)?.removeView(bannerView)
        showRunnable?.let { uiHandler.removeCallbacks(it) }
    }

    private fun cacheAdIfNeeded() {
        if (!Appodeal.isAutoCacheEnabled(Appodeal.BANNER_VIEW)) {
            reactContext.currentActivity?.let { activity ->
                Appodeal.cache(activity, Appodeal.BANNER_VIEW)
            }
        }
    }

    override fun handleEvent(event: String, params: WritableMap?) {
        when (event) {
            BannerEvents.ON_BANNER_LOADED -> dispatchFabricEvent(id, BannerEvents.ON_AD_LOADED, params)
            BannerEvents.ON_BANNER_FAILED_TO_LOAD -> dispatchFabricEvent(id, BannerEvents.ON_AD_FAILED_TO_LOAD, params)
            BannerEvents.ON_BANNER_CLICKED -> dispatchFabricEvent(id, BannerEvents.ON_AD_CLICKED, params)
            BannerEvents.ON_BANNER_EXPIRED -> dispatchFabricEvent(id, BannerEvents.ON_AD_EXPIRED, params)
            else -> Unit
        }
    }

    private fun dispatchFabricEvent(
        viewId: Int,
        eventName: String,
        params: WritableMap?
    ) {
        val dispatcher: EventDispatcher? =
            UIManagerHelper.getEventDispatcherForReactTag(reactContext, viewId)
        dispatcher?.dispatchEvent(OnViewEvent(surfaceId, viewId, eventName, params))
    }

    private class OnViewEvent(
        surfaceId: Int,
        viewId: Int,
        private val eventName: String,
        private val payload: WritableMap?
    ) : Event<OnViewEvent>(surfaceId, viewId) {
        override fun getEventName(): String = eventName
        override fun getEventData(): WritableMap? = payload
    }

    private val reactContext: ReactContext
        get() = context as ReactContext

}

private fun dp2px(dp: Int, dm: DisplayMetrics): Int =
    TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, dp.toFloat(), dm).roundToInt()